### PR TITLE
Set as non-hoistable if a mutable variable is passed to a function in a loop.  Resolves #3699.

### LINKED ISSUE #3699: prange segfault, invalid loop invariant detection
## Reporting a bug

- [x] I am using the latest released version of Numba (most recent is visible in
 the change log (https://github.com/numba/numba/blob/master/CHANGE_LOG).
- [x] I have included below a minimal working reproducer (if you are unsure how
 to write one see http://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports).

The following code fails with `interrupted by signal 11: SIGSEGV` but runs fine if `Parallel=False` (Ubuntu 18.04, Python 3.6)

```python

import random

from numba import njit, prange


@njit
def pick_card(picked: list) -> int:
    a = random.randint(0, 51)
    if a in picked:
        return pick_card(picked)
    else:
        picked.append(a)
    return a


@njit(parallel=True)
def calc(total):
    cards = []
    for suit in range(4):
        for card in range(13):
            cards.append(card)

    for i in prange(total):
        hand = []
        picked = [100]
        for _ in range(13):
            a = pick_card(picked)
            hand.append(cards[a])
    return


calc(100000)
print("Finished!")```

### PR DESCRIPTION
**Pull request by DrTodd13**
Resolves #3699.

### CODE REVIEW

File: numba/npyufunc/parfor.py [hunk 1]
```diff
@@ -660,6 +660,21 @@ def redarraytype_to_sig(redarraytyp):
...
 
+def legalize_names_with_typemap(names, typemap):
+    """ We use ir_utils.legalize_names to replace internal IR variables that have things like
+        periods in them that aren't legal parameter names and we replace those with things like
+        underscores which are legal.  The original variable names are in the typemap so we
+        need to add the legalized name to the typemap as well.
+    """
+    outdict = legalize_names(names)
+    # For each pair in the dict of legalized names...
+    for x,y in outdict.items():
```
Comment (stuartarchibald) [inline numba/npyufunc/parfor.py]: PEP8 space between `x,` and `y`

File: numba/npyufunc/parfor.py [hunk 2]
```diff
@@ -660,6 +660,21 @@ def redarraytype_to_sig(redarraytyp):
     assert isinstance(redarraytyp, types.npytypes.Array)
     return types.npytypes.Array(redarraytyp.dtype, max(1, redarraytyp.ndim - 1), redarraytyp.layout)
 
+def legalize_names_with_typemap(names, typemap):
+    """ We use ir_utils.legalize_names to replace internal IR variables that have things like
```
Comment (stuartarchibald) [inline numba/npyufunc/parfor.py]: I think I understand this sentence but it was quite hard to parse. Is the intent:
> We use ir_utils.legalize_names to replace internal IR variable names containing illegal characters (e.g. period) with a legal character (underscore) so as to create legal variable names.

?

File: numba/npyufunc/parfor.py [hunk 3]
```diff
@@ -513,8 +513,23 @@ def compute_def_once_block(block, def_once, def_more, getattr_taken):
...
+            # therefore can't be hoisted.
+            # For each argument to the function...
+            for argvar in rhs.args:
+                # Get the argument's type.
+                if isinstance(argvar, ir.Var):
+                    argvar = argvar.name
+                avtype = typemap[argvar]
+                # If that type doesn't have a mutable attribute or it does and it's set to
+                # not mutable then this usage is safe for hoisting.
+                if hasattr(avtype, 'mutable') and avtype.mutable:
```
Comment (stuartarchibald) [inline numba/npyufunc/parfor.py]: Is this the same as:
```
if getattr(avtype, 'mutable', False):
```
?

File: numba/npyufunc/parfor.py (hunk 1 above)
Comment (DrTodd13) [reply numba/npyufunc/parfor.py]: Done.

File: numba/npyufunc/parfor.py (hunk 2 above)
Comment (DrTodd13) [reply numba/npyufunc/parfor.py]: Done.

File: numba/npyufunc/parfor.py (hunk 3 above)
Comment (DrTodd13) [reply numba/npyufunc/parfor.py]: Thanks.  I'll remember the idiom one of these days.

File: numba/tests/test_parfors.py [hunk 4]
```diff
@@ -2039,6 +2039,20 @@ def test_impl(n):
...
+            def list_check(X):
+                a = X[-1] + 1
+                if a not in X:
+                    X.append(a)
+                return a
+            for i in range(n):
+                X = [100]
+                a = list_check(X)
+            return a
+        self.prange_tester(test_impl, 10)
```
Comment (stuartarchibald) [inline numba/tests/test_parfors.py]: I think this compiles with or without the fix and it seems like `X` isn't hoisted irrespective of the fix being present, I guess because `list_check` is an inlinable function it's inlined at the call site by the closure inlining pass so there's something that prevents hoisting now present? The function also has a race condition, but I don't think that matters. The parfors diagnostic metadata may help in writing a test:
```python
from numba import prange, njit

@njit
def list_check(X):
        a = X[-1] + 1
        if a not in X:
            X.append(a)
        return a

@njit(parallel=True)
def test_impl(n):
    for i in prange(n):
        X = [100]
        a = list_check(X)
    return a

test_impl(10)

diagnostics = test_impl.get_metadata()[test_impl.signatures[0]]['parfor_diagnostics']
diagnostics.dump(4)

for loop, hoist_info in diagnostics.hoist_info.items():
    for h, stmts in hoist_info.items():
        print("Info %s" % h)
        for stmt in stmts:
            print("loop %s. %s" % (loop, stmt))
        print()
```

Comment (DrTodd13) [reply numba/tests/test_parfors.py]: I changed list_check slightly so that it would definitely fail if X = [100] were hoisted.  I also confirmed manually with parfor diagnostics that it isn't hoisted here.  I switched this test to manual style instead of prange_tester.

**Comment by stuartarchibald** [comment]
Thanks for the fixes. Looks good.
